### ISSUE TITLE
Exception while writing objects, part 2: Electric Boogaloo

### ISSUE THREAD

**Issue by galbacarys** [anishathalye/git-remote-dropbox#36]
Ref. #26 

I'm having the same issue (circa end of April 2016). I initially figured it was the dreaded InsecurePlatformWarning issue (i.e. no SSL context available), but I upgraded all my security related packages and that didn't fix it. I haven't done any debugging but I wanted to let you know that this problem exists for me (at least in the latest Linux Mint x64) on Python 2.7.6.

**Comment by anishathalye** [comment]
Uh oh. Well, thanks for reporting this! If you can give instructions on how to reproduce it, that would be awesome.

**Comment by galbacarys** [comment]
Steps to reproduce are the same as in #26.

I added the `dropbox://` remote, tried to push, and see that files are being written to my dropbox folder. But then the connection hangs, and I get the `Exception while writing objects` issue.

I don't have a trace available, but I can post one tonight once I'm back in front of my normal computer.

**Comment by anishathalye** [comment]
might also be related to #35 

I probably don't have time to spend too much time on this until 5/12, but if anyone else wants to start looking into it before that (and maybe submit a PR), that would be awesome

**Comment by ZachariBarnes** [comment]
Also experiencing this issue, trying to push to an empty folder in a shared directory.
error: exception while writing objects
error: failed to push some refs to 'Dropbox://Ligaments_Inc/Master'

**Comment by anishathalye** [comment]
Welp. Sorry, error reporting is still pretty bad (see #9). If you don't mind editing the code a little bit and trying again so we can get more details about what's going wrong, you can try [this](https://github.com/anishathalye/git-remote-dropbox/issues/15#issuecomment-135314527) (except it's [line 727](https://github.com/anishathalye/git-remote-dropbox/blob/master/git_remote_dropbox/__init__.py#L727) now).

**Comment by arpit1997** [comment]
I got the same kind of error while running the application. I configured my credentials and tried to backup. But it thrown the same error as @galbacarys ` Exception while writing the objects ` 
I am using ubuntu 14.04 with python3.4. The folder which i tried to backup is my django project local repo. It contains some python files and one `db.sqlite3` file. The error log does not show much information so it will be so much helpful if you can improve this @anishathalye 
:micronesia:  :smile:

**Comment by Drenn1** [comment]
Hi, original issue reporter here. I came back for more punishment.

The issue persists on my Arch Linux computers. The output might be more detailed than it was before, though.

It gives me the "Writing objects: X/Y" line. In my repository there are about 5000 objects. When I run the command "git push dropbox master", it gets to maybe object 20 if I'm lucky before it gives me an error. The strange thing, is that this takes several seconds. Each object is taking way too long to push.

```
Writing objects:   0% (19/5303)error: exception while writing objects
error: failed to push some refs to 'dropbox://git/ages-disasm'
```

I've now tried this on my Windows virtual machine in cygwin. It made significantly more progress... however, the slowness was still an issue, and eventually, it failed too.

```
$ time git push dropbox master
Writing objects:  21% (1094/5180)error: exception while writing objects
error: failed to push some refs to 'dropbox://git/ages-disasm'

real    6m50.039s
user    0m29.741s
sys     2m1.662s
```

So, maybe the connection is timing out.

Want to here something REALLY bizarre? Here's what happened after tethering my computer's internet to my phone. (On linux this time, not cygwin)

```
Writing objects:  31% (1657/5303)^C

real    7m44.961s
user    0m0.003s
sys     0m0.000s
```

It got stuck at 31%. It took several minutes to get to that point, and I terminated it a few minutes later.

My home internet isn't great - less than 8 megabits per second. My phone is maybe twice as good. Despite the low transfer rate due to the slowness (about 25-40 kilobytes/second being uploaded in that last case), my phone's better speed seems to have helped. Unless there's some other factor on that network...

**Comment by anishathalye** [comment]
Okay I know my error messages are terrible (#9). I think it would help if we could see exactly what's going wrong.

If you could edit right after line 463 so it looks like this:

```
        except Exception:
            import traceback
            traceback.print_exc()
            self._fatal('exception while writing objects')
```

Then we could see a more detailed error message (and traceback) that could help debug this.

**Comment by Drenn1** [comment]
Here's mine. (Version 0.2.2, Dropbox sdk version 6.9.0)

```
Writing objects:   0% (15/5303)Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/urllib3/connection.py", line 138, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/usr/lib/python3.5/site-packages/urllib3/util/connection.py", line 75, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/urllib3/connectionpool.py", line 594, in urlopen
    chunked=chunked)
  File "/usr/lib/python3.5/site-packages/urllib3/connectionpool.py", line 350, in _make_request
    self._validate_conn(conn)
  File "/usr/lib/python3.5/site-packages/urllib3/connectionpool.py", line 833, in _validate_conn
    conn.connect()
  File "/usr/lib/python3.5/site-packages/urllib3/connection.py", line 275, in connect
    conn = self._new_conn()
  File "/usr/lib/python3.5/site-packages/urllib3/connection.py", line 147, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
requests.packages.urllib3.exceptions.NewConnectionError: <requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8182f12b0>: Failed to establish a new connection: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/requests/adapters.py", line 423, in send
    timeout=timeout
  File "/usr/lib/python3.5/site-packages/urllib3/connectionpool.py", line 643, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/lib/python3.5/site-packages/urllib3/util/retry.py", line 303, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
requests.packages.urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='content.dropboxapi.com', port=443): Max retries exceeded with url: /2/files/upload (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8182f12b0>: Failed to establish a new connection: [Errno -2] Name or service not known',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/git_remote_dropbox/__init__.py", line 453, in _push
    for done, _ in enumerate(res, 1):
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 695, in next
    raise value
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.5/site-packages/git_remote_dropbox/__init__.py", line 291, in __call__
    return method(self._obj, *args)
  File "/usr/lib/python3.5/site-packages/git_remote_dropbox/__init__.py", line 515, in _put_object
    self._connection().files_upload(data, path, mode, mute=True)
  File "/usr/lib/python3.5/site-packages/dropbox/base.py", line 1225, in files_upload
    f,
  File "/usr/lib/python3.5/site-packages/dropbox/dropbox.py", line 249, in request
    timeout=timeout)
  File "/usr/lib/python3.5/site-packages/dropbox/dropbox.py", line 341, in request_json_string_with_retry
    timeout=timeout)
  File "/usr/lib/python3.5/site-packages/dropbox/dropbox.py", line 417, in request_json_string
    timeout=timeout,
  File "/usr/lib/python3.5/site-packages/requests/sessions.py", line 522, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/lib/python3.5/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.5/site-packages/requests/sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3.5/site-packages/requests/adapters.py", line 487, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='content.dropboxapi.com', port=443): Max retries exceeded with url: /2/files/upload (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8182f12b0>: Failed to establish a new connection: [Errno -2] Name or service not known',))
error: exception while writing objects
error: failed to push some refs to 'dropbox://git/ages-disasm'
```

**Comment by anishathalye** [comment]
Huh, looks like an issue with the [Python API for Dropbox](https://github.com/dropbox/dropbox-sdk-python).

**Comment by Drenn1** [comment]
Yeah, that's what you concluded last time unfortunately. May I ask what your internet speed is like? Since the issue was partially alleviated by switching to a faster network, I'm curious whether internet speed is one of the causes. Or rather, whether a good internet speed works around whatever the issue is. If so, perhaps limiting the program's speed with something like netbrake would allow you to debug the issue.

(btw, can you confirm that you saw my initial post, since we replied here at almost the same time? =P)

**Comment by anishathalye** [comment]
I've used git-remote-dropbox at my house, where my internet speed is something like 8 Mbit/s down and 768 Kbit/s up. I've also used it at school where I have 1 Gbit/s up and down 😄

I haven't had this particular issue on either network (and that's what makes it hard for me to debug it).

And yeah, my comment (https://github.com/anishathalye/git-remote-dropbox/issues/36#issuecomment-256510448) was a response to your comment (https://github.com/anishathalye/git-remote-dropbox/issues/36#issuecomment-256507648), if that's what you were asking.

**Comment by Drenn1** [comment]
I was referring to the comment before that (https://github.com/anishathalye/git-remote-dropbox/issues/36#issuecomment-256504424) but ok.

I think I've eliminated the exception by removing the multithreading. The question remains why it's so unbearably slow.

**Comment by anishathalye** [comment]
Yeah, I had seen that comment too. This one (https://github.com/anishathalye/git-remote-dropbox/issues/36#issuecomment-256505474) was written as a response to your earlier comment.

Oh wow, the multithreading causes the exception? Weird. :/

Yeah, see #34. I have no idea why it's slow. I think it's just an issue with the Dropbox API for Python... probably worth doing some testing at some point to figure this out.

**Comment by Drenn1** [comment]
How slow is it for you? My current push operation feels like it could easily take another hour to complete. It's operating at like 3KiB/second upload speed, with occasional spikes to 12KiB/second.

**Comment by anishathalye** [comment]
Oh, mine isn't nearly that slow. Trying it out with this repo (which is admittedly pretty small):

```
$ time git clone dropbox://repos/git-remote-dropbox
Cloning into 'git-remote-dropbox'...
Receiving objects: 100% (97/97), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Receiving objects: 100% (1/1), done.
Checking connectivity... done.
warning: remote HEAD refers to nonexistent ref, unable to checkout.

hub clone dropbox://repos/git-remote-dropbox  3.15s user 2.35s system 23% cpu 23.769 total
$ cd git-remote-dropbox
$ git remote set-url origin dropbox://repos/git-remote-dropbox-test
$ git checkout master
Branch master set up to track remote branch master from origin.
Already on 'master'
$ time git push -u origin master
Writing objects: 100% (97/97), done.
To dropbox://repos/git-remote-dropbox-test
 * [new branch]      master -> master
Branch master set up to track remote branch master from origin.
hub push -u origin master  3.23s user 1.87s system 21% cpu 24.115 total
```

**Comment by Drenn1** [comment]
Welp, it gave me an exception at 93%. So much for that.

**Comment by anishathalye** [comment]
Oh no! :/

Not sure what's wrong…

**Comment by 40fixer** [comment]
I managed to solve the issue by the following:

```diff
diff --git a/git_remote_dropbox/__init__.py b/git_remote_dropbox/__init__.py
index f553b93..d17b78a 100755
--- a/git_remote_dropbox/__init__.py
+++ b/git_remote_dropbox/__init__.py
@@ -343,7 +343,7 @@ class Helper(object):
         """
         # we use fresh connection objects for every use so that multiple
         # threads can have connections simultaneously
-        return dropbox.Dropbox(self._token)
+        return dropbox.Dropbox(self._token, timeout=1000)
 
     def run(self):
         """

```

**Comment by anishathalye** [comment]
Wow, interesting! Do you know what the default timeout is? Also, what's that new timeout? 1 second or 1000 seconds?

**Comment by 40fixer** [comment]
The default was 30 seconds, and the new is 1000 seconds. Probably 1000 is too much, but I just wanted to be sure that this is the issue.

**Comment by astoeckley** [comment]
I was having the same experience. Here is my feedback for what it is worth.

I was pushing my .emacs.d directory for first time, which has a lots of files. The first time I got this:

```
Writing objects:   8% (130/1625)error: exception while writing objects
error: failed to push some refs to 'dropbox:......
```
It got to 8% after about 1 or 2 minutes.

I did it again, and it got further, 139/1625 or 9% before doing the same.

I did it again, and it went to 62% and then it sat there for about 4 minutes doing nothing. The dropbox client icon in my Mac menu bar also stopped updating (it was previously updating with all in-progress pushes, despite that my dropbox folder is not on the local hard drive, turned off via selective sync).

After it paused for 4 minutes, it then completed successfully.

I also immediately pushed the same repo to bitbucket, a first-time push, and it worked quickly in about 5 seconds.

**Comment by astoeckley** [comment]
As a followup question, are there any risks to the integrity of either the local or the remote repositories when a push fails like this with an exception? Are only some of the git artifacats getting uploaded to Dropbox, or is it an all-or-nothing event?

**Comment by anishathalye** [comment]
Weird. I think that's a problem with the Dropbox Python SDK.

Also, pushes to Bitbucket are a lot faster because it uses a different protocol - it doesn't push every object individually like git-remote-dropbox has to (Bitbucket has server-side support that makes this efficient; Dropbox doesn't have anything special for this).

In any case, there is **no risk** of either local or remote repository being damaged. See the [design doc](https://github.com/anishathalye/git-remote-dropbox/blob/master/DESIGN.rst#refs) for an explanation of how this works, if you're interested.

**Comment by anishathalye** [comment]
I think this is fixed by #50. If not, please reopen this issue.

Also, if stuff is still broken, try running with `-v` and see if it gives a stack trace that helps us debug this.
